xds/src/main/java/io/grpc/xds/LocalityStore.java

```diff
@@ -497,5 +443,174 @@ protected Helper delegate() {
       }
     }
+
+    private final class PriorityManager {
+
+      private final List<List<XdsLocality>> priorityTable = new ArrayList<>();
+      private Map<XdsLocality, LocalityInfo> localityInfoMap = ImmutableMap.of();
+      private int currentPriority = -1;
+      private ScheduledHandle failOverTimer;
+
+      /**
+       * Updates the priority ordering of localities with the given collection of localities.
+       * Recomputes the current ready localities to be used.
+       */
+      void updateLocalities(Map<XdsLocality, LocalityInfo> localityInfoMap) {
+        this.localityInfoMap = ImmutableMap.copyOf(localityInfoMap);
+        priorityTable.clear();
+        for (XdsLocality newLocality : localityInfoMap.keySet()) {
+          int priority = localityInfoMap.get(newLocality).priority;
+          while (priorityTable.size() <= priority) {
+            priorityTable.add(new ArrayList<XdsLocality>());
+          }
+          priorityTable.get(priority).add(newLocality);
+        }
+
+        if (currentPriority >= priorityTable.size()) {
+          currentPriority = priorityTable.size() - 1;
+        }
+
+        if (currentPriority == -1) {
+          failOver();
+          return;
+        }
+
+        for (int p = 0; p < priorityTable.size(); p++) {
+          updatePriorityState(p);
+        }
+      }
+
+      /**
+       * Refreshes the group of localities with the given priority. Recomputes the current ready
+       * localities to be used.
+       */
+      private void updatePriorityState(int priority) {
+        if (priority > currentPriority) {
+          return;
+        }
+        List<WeightedChildPicker> childPickers = new ArrayList<>();
+
+        ConnectivityState overallState = null;
+        for (XdsLocality l : priorityTable.get(priority)) {
+          if (!localityMap.containsKey(l)) {
+            initLocality(l);
+          }
+          LocalityLbInfo localityLbInfo = localityMap.get(l);
+          localityLbInfo.reactivate();
+          ConnectivityState childState = localityLbInfo.childHelper.currentChildState;
+          SubchannelPicker childPicker = localityLbInfo.childHelper.currentChildPicker;
+
+          overallState = aggregateState(overallState, childState);
+
+          if (READY == childState) {
+            childPickers.add(
+                new WeightedChildPicker(localityInfoMap.get(l).localityWeight, childPicker));
+          }
+        }
+
+        if (priority == currentPriority || overallState == READY) {
+          updatePicker(overallState, childPickers);
+        }
+
+        if (overallState == READY || overallState == TRANSIENT_FAILURE) {
+          cancelFailOverTimer();
+        }
+
+        if (priority == currentPriority && overallState != READY) {
+          failOver();
+        } else if (priority <= currentPriority && overallState == READY) {
+          goToPriority(priority);
+        }
+      }
+
+      void onLocalityStateUpdate(XdsLocality locality) {
```

Comment:
A slightly better approach is to introduce a method in `PriorityManager`:

```java
// Returns the priority of the given locality or -1 if it is untracked.
int getPriority(XdsLocality locality);
```

Then you don't need to have this `PriorityManager#onLocalityStateUpdate(...)`, which is a bit awkward to be there.

Reply:
It is awkward for ChildHelper in another way:
- You would call `priorityManager` twice, `getPriority()` and `updatePriorityState(priority)`.
- You would check a magic number
  ```
    if (priority >= 0) {
      ...
    }
  ```